### PR TITLE
Doc add storage adapter

### PR DESCRIPTION
**Pull request by lohanidamodar**
## What does this PR do?

Adds Doc for adding new storage adapters.

## Test Plan

N/A

## Related PRs and Issues

(If this PR is related to any other PR or resolves any issue or related to any issue link all related PR and issues here.)

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

YES

### CODE REVIEW

**Comment by lohanidamodar** [comment]
@eldadfux  this is dependent on
#867 .

File: docs/tutorials/add-storage-adapter.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Adding a New Storage Adapter
+
+This document is part of the Appwrite contributors' guide. Before you continue reading this document make sure you have read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/master/CODE_OF_CONDUCT.md) and the [Contributing Guide](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md).
+
+## Getting Started
+
+### Agenda
+
+Storage providers help us use various storage services to store our Appwrite data. As of the writing of these lines we already support Local storage, AWS S3 storage and Digitalocean Spaces Storage.
```
Comment (TorstenDittmann) [inline docs/tutorials/add-storage-adapter.md:9]: Maybe create a link for `Digitalocean Spaces Storage` & co 👍🏻

File: docs/tutorials/add-storage-adapter.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+Note that the class name should start with a capital letter as PHP FIG standards suggest.
+
+Always use properly named environment variables if any credentials are required.
+
+### Introduce new device constant
+Introduce newly added device constant in `src/Storage/Storage.php` alongside existing device constants. The device constant should start with `const DEVICE_<name of device>` as the existing ones.
+
+### Introduce new device tests
+Add test for the newly added device adapter inside `tests/Storage/Device`. Use the tests existing adapters as a reference. The test file and class should be properly named `<Adapter class name>Test.php` and class should be `<Adapter class name>Test`
```
Comment (TorstenDittmann) [inline docs/tutorials/add-storage-adapter.md:27]: Add tests*
Use the existing adapter tests as a reference. **

File: docs/tutorials/add-storage-adapter.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+### Run and verify tests
+Run tests using `vendor/bin/phpunit --configuration phpunit.xml` and verify that everything is working correctly.
+
+If everything goes well, create a new pull request in [utopia-php/storage](https://github.com/utopia-php/storage) library.
+
+### Phase 2
+In this phase we will add support to new storage adapter in Appwrite.
```
Comment (TorstenDittmann) [inline docs/tutorials/add-storage-adapter.md:35]: to the new*

File: docs/tutorials/add-storage-adapter.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+* Note for this to happen, your PR in the first phase should have been merged and new version of [utopia-php/storage](https://github.com/utopia-php/storage) library released.
+
+### Upgrade the utopia-php/storage dependency
+Upgrade the utopia-php/sotrage dependency in `composer.json` file.
+
+### Introduce new environment variables
+If required for the new adapter, may be for credentials, introduce new environment variables. The storage envorinment variables are prefixed as `_APP_STORAGE_DEVICE`. Please read [Adding Environment Variable]() guidelines in order to properly introduce new environment variables.
```
Comment (TorstenDittmann) [inline docs/tutorials/add-storage-adapter.md:43]: environment variables *

File: docs/tutorials/add-storage-adapter.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+### Implement the device case
+In `app/controllers/shared/api.php` inside init function, there is a `switch/case` statements for each supported storage device. Implement the instiation of your device type for your device case. The device cases are the devices constants listed in the `uptopa-php/storage/Storage` class.
```
Comment (TorstenDittmann) [inline docs/tutorials/add-storage-adapter.md:46]: instantiation *

File: docs/tutorials/add-storage-adapter.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+### Test and verify everything works
+In order to test you can switch to your new added device using `_APP_STORAGE_DEVICE` environment variable. Then run `docker-compose build && docker-compose up -d` in order to build the containers with updated changes. Once the containers are running, login to Appwrite console and create a project. Then in storage section, try to upload, preview, delete files.
```
Comment (TorstenDittmann) [inline docs/tutorials/add-storage-adapter.md:49]: `In order to test` = `To test, `

newly added device *

File: docs/tutorials/add-storage-adapter.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+Storage providers help us use various storage services to store our Appwrite data. As of the writing of these lines we already support Local storage, AWS S3 storage and Digitalocean Spaces Storage.
+
+As the storage library is separated into [utopia-php/storage](https://github.com/utopia-php/storage), adding new storage adapter will consist of two phases. First adding and implementing the new adapter in the [utopia-php/storage](https://github.com/utopia-php/storage) and then adding support to new storage adapter in Appwrite.
+
+### Phase 1
+In phase 1, we will introduce and implement the new device adapter in [utopia-php/storage](https://github.com/utopia-php/storage) library.
+
+### Add new adapter
+Add new storage inside `src/Storage/Device/` folder. Use one of the existing ones as a reference. The new adapter class should extend `Device` class and implement all the required methods.
```
Comment (TorstenDittmann) [inline docs/tutorials/add-storage-adapter.md:17]: Add a new storage *

**Review by TorstenDittmann — CHANGES_REQUESTED** [review]

**Review by TorstenDittmann — APPROVED** [review]
lgtm 👍🏻

**Pull request merged**